### ISSUE TITLE
Per test formatteable custom path for per attachment.

### ISSUE THREAD

**Issue by vpenades** [nunit/nunit#4020]
Motivation:
I have lots of tests that output lots of files, and more often than not, I struggle with the somewhat limited file attachment API.

Typically within a test, you write a file using the current directory, or using `TestContext.WorkDirectory` , but if you have many files, or test cases, you need to be careful with file names so they're not overwritten between tests. For this it is possible to build a per test path using `TestContext.Test.ID` and other per test properties.

But in general, this leads to very verbose code.

So I would like to propose these API improvements:
- An attribute to define a formatting that will be used to create an output path, per test
- An AttachmentDirectory property in TestContext.

An example of how it would be used:

```c#
[AttachmentPathFormat("{WorkDirectory}/{ClassName}-Results/{ID}")]
public class TestClass
{
    [Test]
    public void Test()
    {
        // at this point, AttachmentDirectory is ". . . bin/debug/net5.0/Test-Results/0-1003/"
        var thisTestAttachDir = TestContext.AttachmentDirectory;
        var path = Path.Combine(thisTestAttachDir , "hello.txt");       

        System.IO.File.WriteAllText(path, "hello world");
        TestContext.AddTestAttachment(path);
    }
}

```

I have a project that already does pretty much that using extensions, but it's not as nice as it could be if integrated directly.

If there's interest I could submit a PR

**Comment by stevenaw** [comment]
Interesting, thanks @vpenades ! I've labeled this as an `idea` to see if there's general interest.

**Comment by stevenaw** [comment]
I'm closing this as there hasn't been much activity or broader interest in adding this to the framework.

@vpenades Thank you for sharing your solution and extension library. Note that there issue you linked, #548 , was fixed in NUnit 4.2 which may allow for a more built-in way to access what you need to there.
